### PR TITLE
JW-150 Handle DeleteNetwork

### PR DESCRIPTION
**Pull request by m-kostrzewa**
This is a pretty weird handler. When we get DeleteNetwork request, we cannot inspect docker network to retreive tenant and network name from its metadata, because it's already gone. The solution is to compare docker networks that are left with all HNS networks that were created by us and find the diff (there will always be one HNS network more than docker networks). Then we remove the diff network from HNS. Of course, we don't touch the Contrail network.

I made some small refactors as well.

### CODE REVIEW

File: controller/controller.go [hunk 1]
```diff
@@ -177,3 +179,46 @@ func (c *Controller) GetOrCreateInstanceIp(net *types.VirtualNetwork,
 	}
 	return allocatedIP, nil
 }
+
+func (c *Controller) DeleteElementRecursive(parent contrail.IObject) error {
+	log.Debugln("Deleting", parent.GetType(), parent.GetUuid())
+	for err := c.ApiClient.Delete(parent); err != nil; err = c.ApiClient.Delete(parent) {
+		if strings.Contains(err.Error(), "404 Resource") {
+			break
```
Comment (mmasztalerczuk) [inline controller/controller.go]: No log? Not any action?

File: controller/controller.go [hunk 2]
```diff
@@ -177,3 +179,46 @@ func (c *Controller) GetOrCreateInstanceIp(net *types.VirtualNetwork,
...
+				"([a-z-]*\\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})")
+			if err != nil {
+				return err
+			}
+			found := re.FindAll([]byte(msg), -1)
+
+			for _, f := range found {
+				split := strings.Split(string(f), "/")
+				typename := split[0]
+				UUID := split[1]
```
Comment (mmasztalerczuk) [inline controller/controller.go:208]: Can split have only one item?

File: driver/driver.go [hunk 3]
```diff
@@ -369,26 +398,75 @@ func (d *ContrailDriver) RevokeExternalConnectivity(req *network.RevokeExternalC
...
-	return tenant, netName, nil
+	return &meta, nil
+}
+
+func (d *ContrailDriver) dockerNetworksMeta() ([]NetworkMeta, error) {
+	var meta []NetworkMeta
+
+	docker, err := dockerClient.NewEnvClient()
+	if err != nil {
+		return meta, err
```
Comment (mmasztalerczuk) [inline driver/driver.go]: meta? Not nil?

File: driver/driver.go [hunk 4]
```diff
@@ -369,26 +398,75 @@ func (d *ContrailDriver) RevokeExternalConnectivity(req *network.RevokeExternalC
...
+func (d *ContrailDriver) hnsNetworksMeta() ([]NetworkMeta, error) {
+	hnsNetworks, err := d.hnsMgr.ListNetworks()
+	if err != nil {
+		return nil, err
+	}
+
+	var meta []NetworkMeta
+	for _, net := range hnsNetworks {
+		splitName := strings.Split(net.Name, ":")
+		tenantName := splitName[1]
```
Comment (mmasztalerczuk) [inline driver/driver.go]: What if split has length one?

**Review by mmasztalerczuk — APPROVED** [review]
Some question, but nothing important.

File: controller/controller.go (hunk 2 above)
Comment (m-kostrzewa) [reply controller/controller.go:208]: Regex finds all strings like "`virtual-network/23e300f4-ab1a-4d97-a1d9-9ed69b601e17`", so the split  around "/" must contain both

File: driver/driver.go (hunk 4 above)
Comment (m-kostrzewa) [reply driver/driver.go]: hnsMgr.ListNetworks() already sanitizes this. Added comment in code.

File: driver/driver.go [hunk 5]
```diff
@@ -175,15 +181,38 @@ func (d *ContrailDriver) DeleteNetwork(req *network.DeleteNetworkRequest) error
...
 		return err
 	}
-	return nil
+
+	var toRemove *NetworkMeta
+	toRemove = nil
+	for _, hnsMeta := range hnsNetsMeta {
+		matchFound := false
+		for _, dockerMeta := range dockerNetsMeta {
+			if dockerMeta.tenant == hnsMeta.tenant && dockerMeta.network == hnsMeta.network {
```
Comment (mclapinski) [inline driver/driver.go:203]: Maybe you could write a function for comparison between dockerMeta and hnsMeta

File: driver/driver.go [hunk 6]
```diff
@@ -175,15 +181,38 @@ func (d *ContrailDriver) DeleteNetwork(req *network.DeleteNetworkRequest) error
...
+				matchFound = true
+				break
+			}
+		}
+		if !matchFound {
```
Comment (michal-clapinski) [inline driver/driver.go]: I think it should be "if matchFound"

File: driver/driver.go [hunk 7]
```diff
@@ -202,12 +231,12 @@ func (d *ContrailDriver) CreateEndpoint(req *network.CreateEndpointRequest) (*ne
...
 	}
 
-	tenant, netName, err := d.tenantAndNetnameFromDockerNetwork(req.NetworkID)
+	meta, err := d.networkMetaFromDockerNetwork(req.NetworkID)
 	if err != nil {
 		return nil, err
 	}
 
-	contrailNetwork, err := d.controller.GetNetwork(tenant, netName)
+	contrailNetwork, err := d.controller.GetNetwork(meta.tenant, meta.network)
```
Comment (mclapinski) [inline driver/driver.go:242]: Maybe you could rewrite GetNetwork to handle **meta** instead of **tenant** and **network**

File: driver/driver.go [hunk 8]
```diff
@@ -255,7 +284,7 @@ func (d *ContrailDriver) CreateEndpoint(req *network.CreateEndpointRequest) (*ne
 	// HNS needs MACs like 11-22-AA-BB-CC-DD
 	formattedMac := strings.Replace(strings.ToUpper(contrailMac), ":", "-", -1)
 
-	hnsNet, err := d.hnsMgr.GetNetwork(tenant, netName)
+	hnsNet, err := d.hnsMgr.GetNetwork(meta.tenant, meta.network)
```
Comment (mclapinski) [inline driver/driver.go:290]: same comment as above

**Review by mclapinski — CHANGES_REQUESTED** [review]

**Review by mclapinski — APPROVED** [review]

**Pull request merged**